qemu/tests/block_hotplug.py

```diff
@@ -88,153 +75,83 @@ def unplug_device(vm, get_disk_cmd, device):
+                if params.get("os_type") == "windows":
+                    if iteration == 0:
+                        error_context.context("Format disk", logging.info)
+                        utils_misc.format_windows_disk(session, disk_index[num],
+                                                       mountpoint=disk_letter[num])
+                error_context.context("Check block device after hotplug.",
+                                      logging.info)
+                if params.get("disk_op_cmd"):
+                    if params.get("os_type") == "linux":
+                        test_cmd = params.get("disk_op_cmd") % (disk, disk)
+                    elif params.get("os_type") == "windows":
+                        test_cmd = params.get("disk_op_cmd") % (disk_letter[num],
+                                                                disk_letter[num])
+                        test_cmd = utils_misc.set_winutils_letter(session, test_cmd)
                     else:
-                        controller_list.append(controller)
-
-                drive = qdevices.QRHDrive("block%d" % num)
-                drive.set_param("file", find_image(img_list[num + 1]))
-                drive.set_param("format", img_format_type)
-                drive_id = drive.get_param("id")
-                drive.hotplug(vm.monitor)
-
-                device.set_param("drive", drive_id)
-                device.set_param("id", "block%d" % num)
-                if params.get("need_controller", "no") == "yes" and bool(random.randrange(2)):
-                    device.set_param("bus", controller.get_param("id")+'.0')
-                blk_extra_param = params.get("blk_extra_params_%s" % img_list[num + 1])
-                if blk_extra_param:
-                    for item in blk_extra_param.split():
-                        key, value = item.split("=", 1)
-                        device.set_param(key, value)
-                device.hotplug(vm.monitor)
-                ver_out = device.verify_hotplug("", vm.monitor)
-                if not ver_out:
-                    err = "%s is not in qtree after hotplug" % pci_type
-                    test.fail(err)
-                plug_status = utils_misc.wait_for(lambda: len(get_new_disk(disks_before_plug,
-                                                  find_disk(vm, get_disk_cmd))) != 0, pause)
-                if plug_status:
-                    disks_after_plug = find_disk(vm, get_disk_cmd)
-                    new_disks = get_new_disk(disks_before_plug, disks_after_plug)
+                        test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
+
+                    status, output = session.cmd_status_output(test_cmd,
+                                                               timeout=disk_op_timeout)
+                    if status:
+                        test.fail("Check for block device failed ")
+
+                if len(devs) > 2:
+                    devs.pop(1)
```

Comment:
Well currently only max 3 devices are added `hba->drive->device`. Assuming the bus requires some additional device we'd get `bus_device->hba->drive->device`. For now it's theoretical, but I'd expect it to be `-2` device and not `1` device which is the `drive`.

Anyway the solutions I'd suggest are (ordered by my liking):

```python
devs = [dev for dev in devs if not isinstance(dev, QDrive)]
if not devs:
    test.fail
```

```python
if len(devs) == 2:
    pass
elif len(devs) == 3:
    devs.pop(1)
else:
    test.fail("Unsupported number of devices returned: %s" % devs)
```

```python
if len(devs) > 2:
    devs.pop(-2)
elif len(devs) == 2:
    device_list.append(devs[1])
else:
    test.fail
```

Anyway it's just theoretical (that's why raising an exception would be IMO best solution) so please let me know what do you think about this.

Reply:
Prefer your solution 1, so adopt it, thank you for your perfect advice.